### ISSUE TITLE
Finish tests for the Director

### ISSUE THREAD

**Issue by TrentHouliston** [NUbots/NUbots#913]
The director algorithm needs a few more tests to ensure that it has good coverage.
#857 implemented many but there were some that were not finished in that PR.
We need to finish them once the director algorithm is finished

- [x] simple test: test that emitting a task executes
- [ ] pass blocked task: test that one task is blocked other tasks can still use the resources it can't
- [x] priority steal: test that higher priority tasks can steal from lower priority tasks
- [x] start stop: test that the `Start` and `Stop` dsl words run when expected
- [x] triggered provider: test that you can trigger providers using other nuclear bind methods
- [x] needs blocking: test that if you need something and it's not available you're blocked
- [x] needs joining: test that if you need something and it becomes available later the task runs
- [x] needs priority loss: test that if a provider loses one of its dependent needs, it stops running all tasks and is blocked.
- [ ] needs when loss causing: check that if a provider with a needs has one of its needs lost to a when with a causing that it cancels all the other needs and runs that causing
- [ ] needs peeling: test that if you need something and someone else steals it, your entire needs tree gets kicked out
- [x] entire task: test that if you emit several non optional tasks, either all run or none do
- [x] optionals are optional: test that optional tasks don't need to run to run the rest
- [x] done task: test that emitting a `Done` task reruns your parent
- [x] when blocking: test that if a `When` condition isn't met the task is blocked
- [x] when causing: test that if `When` blocked and a `Causing` can provide it, it runs as expected
- [x] when causing ladder: test that when causing relationships can be applied recursively
- [ ] when causing ladder leaves: test that when a causing ladder happens, only leaves at the same level get executed
- [x] when causing multi: test that multiple when/causing relationships can run at once if they are orthogonal
- [ ] when causing stopped: test that when running a causing if the causing provider loses its task you reevaluate
- [ ] when causing new: test that if a previously inactive provider with an appropriate causing starts, the blocked provider will try to push it
